src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
+      // (`pressed == true`), it will continue to be translated into
+      // its corresponding modifier in `beforeReportingState()`.
+      state_[n].active = false;
+      state_[n].sticky = false;
+    } else {
+      // The OneShot key has been pressed while it is active, but not
+      // (yet) sticky. First we determine if it's a candidate for
+      // becoming sticky: it can only become sticky if the same key is
+      // pressed twice in a row, and that key is configured to be
+      // sticky.
+      if ((key_addr == prev_key_addr_) && state_[n].stickable) {
+        // In addition, it must have been "double-tapped": pressed
+        // twice in less than `double_tap_time_out` milliseconds. But
+        // if that variable isn't set (negative), we use the default
+        // `time_out` value instead.
+        uint16_t dtto = (double_tap_time_out < 0) ? time_out : double_tap_time_out;
+        if (! Runtime.hasTimeExpired(start_time_, dtto)) {
+          // Making a OneShot sticky is as simple as setting its
+          // `sticky` flag; the other hook functions handle the rest.
+          state_[n].sticky = true;
+        } else {
+          // If two presses weren't quick enough, this second press
+          // will terminate the OneShot behaviour (see above).
+          state_[n].active = false;
+        }
+      } else {
+        // If this OneShot key is not "stickable", we deactivate
+        // it. It still remains in effect until it is released.
+        state_[n].active = false;
+      }
+    }
+    prev_key_addr_ = key_addr;
+
+  } else if (keyToggledOff(key_state)) {
+    state_[n].pressed = false;
+    if (! state_[n].active) {
+      releaseOneShot(n);
+    }
 
-  if (keyIsPressed(keyState)) {
-    prev_key_ = mapped_key;
-    if (!(mapped_key >= Key_LeftControl && mapped_key <= Key_RightGui) &&
-        !(mapped_key.getFlags() == (KEY_FLAGS | SYNTHETIC | SWITCH_TO_KEYMAP))) {
-      should_cancel_ = true;
+  } else {
+    if (state_[n].active && !state_[n].sticky) {
+      // If a OneShot key is held (on its initial press) long enough
+      // (i.e. > `hold_time_out` milliseconds), deactivate it. It
+      // still remains in effect until it is released.
+      if (Runtime.hasTimeExpired(start_time_, hold_time_out)) {
+        state_[n].active = false;
+      }
     }
   }
 
+  // Finally, we allow event processing to continue so that plugins
+  // later in the processing order can act on OneShot key events (in
+  // particular, ActiveModColor).
   return EventHandlerResult::OK;
 }
 
+
+// For any active OneShot modifier keys, keep those modifiers active
+// in the keyboard HID report.
 EventHandlerResult OneShot::beforeReportingState() {
-  for (uint8_t i = 0; i < ONESHOT_KEY_COUNT / 2; i++) {
-    if (state_[i].active) {
-      activateOneShot(i);
+  for (uint8_t i{0}; i < ONESHOT_KEY_COUNT; ++i) {
+    if (state_[i].active || state_[i].pressed) {
+      sustainOneShot(i);
```

Comment:
Here are two subtle, but important changes. First, I decided to include the OneShot layer keys as well as the modifiers. It's not necessary, but it more closely emulates what happens if the user holds down a normal layer shift key.
Second, when a OneShot key has been cancelled, it's `active` flag will be cleared, but its `pressed` flag will still be set, so I needed to add the second condition.
Third, `activateOneShot()` (both the old and new version) sends a "key toggled on" event, but in this new version, the toggle-on gets sent by the event handler, and `beforeReportingState()` only needs to send the "key held" event, so I'm using the new `sustainOneShot()` function instead.